bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/internal/BidiUtil.java

```diff
@@ -456,66 +456,78 @@ public static void removeLanguageListener (long hwnd) {
  * <p>
  *
- * @param text
- *            Text base direction should be resolved for.
+ * @param text Text base direction should be resolved for.
  * @return SWT#LEFT_RIGHT or SWT#RIGHT_TO_LEFT if the text contains strong
  *         characters and thus the direction can be resolved, SWT#NONE
  *         otherwise.
  * @since 3.105
  */
-public static int resolveTextDirection (String text) {
-	if (text == null) return SWT.NONE;
-	int length = text.length();
-	if (length == 0) return SWT.NONE;
-	char[] rtlProbe = {' ', ' ', '1'};
-	char[] ltrProbe = {'\u202b', 'a', ' '};
-	char[] numberProbe = {'\u05d0', ' ', ' '};
-	GCP_RESULTS result = new GCP_RESULTS();
-	result.lStructSize = GCP_RESULTS.sizeof;
-	int nGlyphs = result.nGlyphs = ltrProbe.length;
-	long hHeap = OS.GetProcessHeap();
-	long lpOrder = result.lpOrder = OS.HeapAlloc(hHeap, OS.HEAP_ZERO_MEMORY, nGlyphs * 4);
-	long hdc = OS.GetDC(0);
-	int[] order = new int[1];
+public static int resolveTextDirection(String text) {
+	if (text == null)
+		return SWT.NONE;
 	int textDirection = SWT.NONE;
-	for (int i = 0; i < length; i++) {
-		char ch = text.charAt(i);
-		rtlProbe[0] = ch;
-		OS.GetCharacterPlacement(hdc, rtlProbe, rtlProbe.length, 0, result, OS.GCP_REORDER);
-		OS.MoveMemory(order, result.lpOrder, 4);
-		if (order[0] == 2) {
-			textDirection = SWT.RIGHT_TO_LEFT;
-			break;
-		}
-		if (textDirection == SWT.LEFT_TO_RIGHT) {
-			// If textDirection is already LTR, skip probing for LTR again.
-			continue;
+	for (int i = 0; i < text.length(); i++) {
+		char c = text.charAt(i);
+		byte directionality = Character.getDirectionality(c);
+		int strongDirection = getStrongDirection(directionality);
+		if (strongDirection != SWT.NONE) {
+			textDirection = strongDirection;
 		}
-		ltrProbe[2] = ch;
-		OS.GetCharacterPlacement(hdc, ltrProbe, ltrProbe.length, 0, result, OS.GCP_REORDER);
-		OS.MoveMemory(order, result.lpOrder + 4, 4);
-		if (order[0] == 1) {
-			numberProbe[2] = ch;
-			OS.GetCharacterPlacement(hdc, numberProbe, numberProbe.length, 0, result, OS.GCP_REORDER);
-			OS.MoveMemory(order, result.lpOrder, 4);
-			if (order[0] == 0) {
-				textDirection = SWT.LEFT_TO_RIGHT;
-				// Do-not break here, instead scan the complete text for any RTL possibility
-			}
+		if (textDirection == SWT.RIGHT_TO_LEFT) {
+			break;
 		}
 	}
-	OS.ReleaseDC (0, hdc);
-	OS.HeapFree(hHeap, 0, lpOrder);
 	return textDirection;
+}
+
+public static int getStrongDirection(byte directionality) {
+	switch (directionality) {
+	// Strong:
+	case Character.DIRECTIONALITY_LEFT_TO_RIGHT:
+		return SWT.LEFT_TO_RIGHT;// Strong bidirectional character type "L" in the Unicode specification.
+	case Character.DIRECTIONALITY_RIGHT_TO_LEFT:
+		return SWT.RIGHT_TO_LEFT;// Strong bidirectional character type "R" in the Unicode specification.
+	case Character.DIRECTIONALITY_RIGHT_TO_LEFT_ARABIC:
+		return SWT.RIGHT_TO_LEFT;// Strong bidirectional character type "AL" in the Unicode specification.
```

Comment:
Less code here, like below:
case Character.DIRECTIONALITY_RIGHT_TO_LEFT:
case Character.DIRECTIONALITY_RIGHT_TO_LEFT_ARABIC:
	return SWT.RIGHT_TO_LEFT;// Strong bidirectional character type "AL" in the Unicode specification.